File: frontend/src/lib/quran/surahRange.ts
/**
 * Mirror of backend/records/services/surah_range.py. Conservative parser:
 * strips verse numbers/ranges, keeps distinct surah names. Used client-side
 * for the evaluation preparation readiness card.
 */

const SEPARATOR_PATTERN = /[,،؛;/|\n\r]+/;
const VERSE_PATTERN = /\d+\s*[-–—]\s*\d+|\d+/g;

export function parseSurahRange(surahRange: string | null | undefined): string[] {
    if (!surahRange) return [];
    const raw = String(surahRange).trim();
    if (!raw) return [];

    const result: string[] = [];
    const seen = new Set<string>();

    for (const part of raw.split(SEPARATOR_PATTERN)) {
        const trimmed = part.trim();
        if (!trimmed) continue;
        const name = trimmed.replace(VERSE_PATTERN, "").trim();
        if (!name) continue;
        if (seen.has(name)) continue;
        seen.add(name);
        result.push(name);
    }

    return result;
}

File: backend/records/services/surah_range.py
"""
Conservative surah-range parser for evaluations.

Accepts inputs like:
  "البقرة 1-40"
  "البقرة"
  "الفاتحة, الناس"
  "البقرة 1-40, آل عمران 1-20"

Returns a list of distinct surah names (verse numbers stripped). If an
input is ambiguous or empty, it is skipped rather than corrupting the
mastery reset downstream.
"""

from __future__ import annotations

import re


_SEPARATORS = re.compile(r"[,،؛;/|\n\r]+")
_VERSE_PATTERN = re.compile(r"\d+\s*[-–—]\s*\d+|\d+")


def surah_range_to_names(surah_range: str | None) -> list[str]:
    if not surah_range:
        return []
    raw = str(surah_range).strip()
    if not raw:
        return []

    result: list[str] = []
    seen: set[str] = set()

    for part in _SEPARATORS.split(raw):
        if not part:
            continue
        name = _VERSE_PATTERN.sub("", part).strip()
        if not name:
            continue
        if name in seen:
            continue
        seen.add(name)
        result.append(name)

    return result
